llama_stack/distribution/datatypes.py

```diff
@@ -64,6 +65,103 @@ class AccessAttributes(BaseModel):
+
+
+class AttributeReference(str, Enum):
+    RESOURCE_ROLES = "resource.roles"
+    RESOURCE_TEAMS = "resource.teams"
+    RESOURCE_PROJECTS = "resource.projects"
+    RESOURCE_NAMESPACES = "resource.namespaces"
+
+
+class Condition(BaseModel):
+    user_in: AttributeReference | list[AttributeReference] | str | None = None
+    user_not_in: AttributeReference | list[AttributeReference] | str | None = None
+
+
+class AccessRule(BaseModel):
+    """Access rule based loosely on cedar policy language
+
+    A rule defines a list of action either to permit or to forbid. It may specify a
+    principal or a resource that must match for the rule to take effect. The resource
+    to match should be specified in the form of a type qualified identifier, e.g.
+    model::my-model or vector_db::some-db, or a wildcard for all resources of a type,
+    e.g. model::*. If the principal or resource are not specified, they will match all
+    requests.
+
+    A rule may also specify a condition, either a 'when' or an 'unless', with additional
+    constraints as to where the rule applies. The constraints at present are whether the
+    user requesting access is in or not in some set. This set can either be a particular
+    set of attributes on the resource e.g. resource.roles or a literal value of some
+    notion of group, e.g. role::admin or namespace::foo.
+
+    Rules are tested in order to find a match. If a match is found, the request is
+    permitted or forbidden depending on the type of rule. If no match is found, the
+    request is denied. If no rules are specified, a rule that allows any action as
+    long as the resource attributes match the user attributes is added
+    (i.e. the previous behaviour is the default).
+
+    Some examples in yaml:
+
+    - permit:
+      principal: user-1
+      actions: [create, read, delete]
+      resource: model::*
+      comment: user-1 has full access to all models
+    - permit:
+      principal: user-2
+      actions: [read]
+      resource: model::model-1
+      comment: user-2 has read access to model-1 only
+    - permit:
+      actions: [read]
+      when:
+        user_in: resource.namespaces
+      comment: any user has read access to any resource with matching attributes
+    - forbid:
+      actions: [create, read, delete]
+      resource: vector_db::*
+      unless:
+        user_in: role::admin
+      comment: only user with admin role can use vector_db resources
+
+    """
+
+    permit: Scope | None = None
+    forbid: Scope | None = None
+    when: Condition | list[Condition] | None = None
+    unless: Condition | list[Condition] | None = None
+    comment: str | None = None
+
+    @model_validator(mode="after")
+    def validate_rule_format(self) -> Self:
+        _require_one_of(self, "permit", "forbid")
+        _mutually_exclusive(self, "permit", "forbid")
+        _mutually_exclusive(self, "when", "unless")
+        return self
+
+
 class ResourceWithACL(Resource):
```

Comment:
what's the future of this thing? are we going to have access attributes stored with resources still?

Reply:
> My suggestion would be to be more explicit that the attributes stored against a resource are those of the user who created it (if it was dynamically created). I would also suggest recording the principal of the creator along with those attributes, and updating the condition on the access rules to allow restriction to the owner/creator.

yes this makes sense to me. I really don't like the current design (which I implemented) of storing the _attributes_ of the user who created the resource. we should

- have policies which say "is this user allowed to create a resource X"
- if yes, we will record the creator 
- then we can have a policy rule which says "only allow read, update if principal == creator"

Reply:
@grs actually I thought about this a bit and it seems like we should probably store as few pieces as possible in the Resource attributes because one would need to invalidate these things also? 

first, I was thinking we could only store principal. and then have a service which can retrieve claims _given_ a principal. then one could write a rules like `permit read if creator principal (from resource) and current principal (auth_user) has an intersecting claim`. there is still the problem now that group ownership can be dynamic (a principal can get evicted from a group or be added to one) so I am not sure this is a good solution either.

Reply:
@ashwinb I have just pushed an additional commit that changes the conditions and the resource attributes.

With this you can define rules that constrain access to only the user who created the object or to users with some attribute (e.g. in the admin role). I think these will be the most common type of restriction.

You can also restrict access to e.g. users in a team that the creator of the resource was in when creating it. While I agree that the groupings a given user belongs to may change, that only applies to this last category of restriction. I think that as long as we are clear that this type of restriction applies to the creators membership at the time of creation, it should be fine.